packages/elements/list/src/transforms/toggleList.ts

```diff
@@ -1,46 +1,139 @@
 import {
   ELEMENT_DEFAULT,
+  findNode,
   getNodes,
+  isCollapsed,
   setNodes,
-  someNode,
   wrapNodes,
 } from '@udecode/plate-common';
 import { getPlatePluginType, SPEditor, TElement } from '@udecode/plate-core';
-import { Editor } from 'slate';
+import { Editor, Node, NodeEntry, Path, Range } from 'slate';
 import { ELEMENT_LI, ELEMENT_LIC } from '../defaults';
+import { getListItemEntry, getListTypes } from '../queries';
 import { unwrapList } from './unwrapList';
 
 export const toggleList = (editor: SPEditor, { type }: { type: string }) => {
-  if (!editor.selection) return;
+  if (!editor.selection) {
+    return;
+  }
 
   Editor.withoutNormalizing(editor, () => {
-    const isActive = someNode(editor, { match: { type } });
+    if (isCollapsed(editor.selection)) {
+      // selection is collapsed
+      const res = getListItemEntry(editor);
 
-    unwrapList(editor);
+      if (res) {
+        const { list } = res;
+        if (list[0].type !== type) {
+          setNodes(
+            editor,
+            { type },
+            {
+              at: editor.selection as Range,
+              match: (n) => getListTypes(editor).includes(n.type),
+              mode: 'lowest',
+            }
+          );
+        } else {
+          unwrapList(editor);
+        }
+      } else {
+        const list = { type, children: [] };
+        wrapNodes(editor, list);
 
-    setNodes<TElement>(editor, {
-      type: getPlatePluginType(editor, ELEMENT_DEFAULT),
-    });
+        const nodes = [
+          ...getNodes(editor, {
+            match: { type: getPlatePluginType(editor, ELEMENT_DEFAULT) },
+          }),
+        ];
+        setNodes(editor, { type: getPlatePluginType(editor, ELEMENT_LIC) });
 
-    if (!isActive) {
-      const list = { type, children: [] };
-      wrapNodes(editor, list);
+        const listItem = {
+          type: getPlatePluginType(editor, ELEMENT_LI),
+          children: [],
+        };
 
-      const nodes = [
-        ...getNodes(editor, {
-          match: { type: getPlatePluginType(editor, ELEMENT_DEFAULT) },
-        }),
-      ];
-      setNodes(editor, { type: getPlatePluginType(editor, ELEMENT_LIC) });
+        for (const [, path] of nodes) {
+          wrapNodes(editor, listItem, {
+            at: path,
+          });
+        }
+      }
+    } else {
+      // selection is a range
+      const [statPoint, endPoint] = Range.edges(editor.selection!);
+
+      const commonEntry: NodeEntry<Node> = Node.common(
```

Comment:
Node is the default.

```suggestion
      const commonEntry: NodeEntry = Node.common(
```